gnocchi/amqp1d.py

```diff
@@ -0,0 +1,251 @@
+        """Based of FORMAT_VL from collectd/src/daemon/common.h.
+
+        The biggest difference is that we don't prepend the host and append the
+        index of the value, and don't use slash.
+
+        """
+        return (value["plugin"] + ("-" + value["plugin_instance"]
+                                   if value["plugin_instance"] else "")
+                + "@"
+                + value["type"] + ("-" + value["type_instance"]
+                                   if value["type_instance"] else "")
+                + "-" + str(index))
+
+    def treat_metric(self, host, metric_name, metric_type,
+                     value):
+        """Collectd.
+
+        Statistics in collectd consist of a value list. A value list includes:
+        Values, can be one of:
+        Derive: used for values where a change in the value since it's last
+        been read is of interest. Can be used to calculate and store a rate.
+        Counter: similar to derive values, but take the possibility of a
+        counter wrap around into consideration.
+        Gauge: used for values that are stored as is.
+        Absolute: used for counters that are reset after reading.
+
+        """
+
+        if metric_type == "absolute":
+            if host not in self.absolute:
+                self.absolute[host] = {}
+            self.absolute[host][metric_name] = incoming.Measure(
+                utils.dt_in_unix_ns(utils.utcnow()), value)
+        elif metric_type == "gauge":
+            if host not in self.gauges:
+                self.gauges[host] = {}
+            self.gauges[host][metric_name] = incoming.Measure(
+                utils.dt_in_unix_ns(utils.utcnow()), value)
+        elif metric_type == "counter" or metric_type == "derive":
+            if host not in self.counters:
+                self.counters[host] = {}
+            self.counters[host][metric_name] = incoming.Measure(
+                utils.dt_in_unix_ns(utils.utcnow()), value)
+        else:
+            raise ValueError("Unknown metric type '%s'" % metric_type)
+
+    def flush(self):
+        for host, measures in itertools.chain(
+            six.iteritems(
+                self.counters), six.iteritems(
+                self.gauges), six.iteritems(
+                self.absolute)):
+
+            host_id = self.conf.amqp1d.resource_name + \
+                ":" + host.replace("/", "_")
+            resources = self.indexer.list_resources(
+                self.conf.amqp1d.resource_name,
+                attribute_filter={"=": {"host": host}})
+            if not resources:
+                self._ensure_resource_exists(host_id, host)
+                resources = self.indexer.list_resources(
+                    self.conf.amqp1d.resource_name,
+                    attribute_filter={"=": {"host": host}})
+
+            resource = self.indexer.get_resource(
+                self.conf.amqp1d.resource_name,
+                resources[0].id, with_metrics=True)
```

Comment:
so you're doing:

- list
- create
- list
- get

OR

- list
- get

just for one resource. That's huge.

Either just do one list for ALL resources and then create the missing one, or, just try to create the resource each time and ignore ResourceAlreadyExists. But there's too many operations right now.